x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pytorch/process/PyTorchStateStreamer.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.pytorch.process;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.get.GetResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.client.OriginSettingClient;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.xpack.core.ml.inference.trainedmodel.pytorch.ModelStorage;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+
+/**
+ * PyTorch models in the TorchScript format are binary files divided
+ * into small chunks and base64 encoded for storage in Elasticsearch.
+ * The model is restored by base64 decoding the stored state and streaming
+ * the binary objects concatenated in order. There is no delineation between
+ * individual chunks the state should appear as one contiguous file.
+ */
+public class PyTorchStateStreamer {
+
+    private static final Logger logger = LogManager.getLogger(PyTorchStateStreamer.class);
+
+    private final OriginSettingClient client;
+    private final ByteBuffer lengthBuffer;
+    private volatile boolean isCancelled;
+
+    public PyTorchStateStreamer(Client client) {
+        this.client = new OriginSettingClient(Objects.requireNonNull(client), ML_ORIGIN);
+        lengthBuffer = ByteBuffer.allocate(4);
+    }
+
+    /**
+     * Cancels the state streaming at the first opportunity.
+     */
+    public void cancel() {
+        isCancelled = true;
+    }
+
+    /**
+     * Firsts writes the size of the model so the native process can
+     * allocated memory then writes the chunks of binary state.
+     *
+     * @param modelStorage  Metadata for the model state
+     * @param restoreStream The stream to write to
+     * @throws IOException
+     */
+    public void writeStateToStream(ModelStorage modelStorage, OutputStream restoreStream) throws IOException {
+
+        writeModelSize(modelStorage.getModelId(), modelStorage.getModelSize(), restoreStream);
+
+        List<String> docIds = modelStorage.documentIds();
+        for (String docId : docIds) {
+            if (isCancelled) {
+                return;
+            }
+
+            GetResponse got = client.prepareGet(modelStorage.getIndex(), docId)
```

Comment:
This should use an IDs search with size 1 across the pattern of indices that might store the state documents.  This is what we do in the `StateStreamer` class: https://github.com/elastic/elasticsearch/blob/a92a647b9f17d1bddf5c707490a19482c273eda3/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/StateStreamer.java#L75-L77

If you look at the history of that class you'll see we started out with using the `Get` API, then switched to the `Search` API with an IDs search later on, because it's more robust to indices being split up, upgraded or reindexed.  So we should use the more robust approach here from the start.

Reply:
The use case is slightly different compared to AD jobs where the model state is updated over time whereas PyTorch models are static. An AD job might start writing index .ml-state-00001 then over time and successive upgrades move to .ml-state-00003. PyTorch models will be put once and if they are moved or the index deleted then the model storage document will have to be updated. Data streams and ILM are designed for fast changing time series data which I didn't consider a suitable paradigm for stable model storage. Upgraded and reindexed indices would be better handled with index aliases. 

There is little difference between an ID search and Get request and the change is simple but so I will make it but for now I'd like to avoid spreading a model over multiple indices a this stage. 

`ModelStorage:: getIndex()` is hard coded for now because there is an ongoing conversation about where to store the models and this fills the gap.

Reply:
> Data streams and ILM are designed for fast changing time series data

The maximum recommended shard size is 50GB, and our internal indices have 1 shard.  ILM (outside of data streams) is also designed for rolling over indices based on size instead of time.

> for now I'd like to avoid spreading a model over multiple indices a this stage

We won't set out to deliberately spread models over multiple indices, but at some point we're going to have to introduce a solution for the problem of the inference index getting too large, and it will be very hard for that solution to reliably keep models in the same index.  For example, if we used size based ILM (tricky as the inference index is a system index) then it could roll over in between the documents of a single model.  If we used some sort of custom rollover due to it being a system index then that code will be incredibly complicated if it has to avoid rolling over part way through indexing the documents of a single model.  And if the eventual solution is to store the models in `.ml-state` then that is already using size-based ILM, and could very easily roll over in between the documents of a single model.

> `ModelStorage:: getIndex()` is hard coded for now because there is an ongoing conversation about where to store the models and this fills the gap

There is no indication in the code that the way it is now is temporary.  Please add a TODO because if it accidentally got left like this when the feature branch is merged to master then for any subsequent problem the person looking into it would be having to use `git blame` to find this PR to find why it was like that.

Reply:
I've reworked this PR significantly the `ModelStorage` class has gone and I'm reusing definitions used in boosted tree model storage. This means the code has to work the same as the existing model loading code in TrainedModelProvider which does support searching for model chunks over multiple indices but has post-processing logic to limit the results to those found in the first index.

https://github.com/elastic/elasticsearch/blob/97f959ed0de4ad2a11a2e3d2a332de7f3547a344/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java#L1081

@benwtrent can you comment on the bit of code linked above please. Was the reason for this that you were worried that model IDs, and therefore document IDs may not be unique over multiple indices?

Reply:
Preamble: I don't think configuration indices should ever be managed by ILM. The current trained model index (inference config) is not managed my ILM. This adds unnecessary complexity. Honestly, if we are THIS concerned around static model configurations causing shards to increase, we really need to come up with another storage mechanism outside the index. The more and more we talk about "large blob storage in an index" the more it gets complicated, hacky, and unmaintainable. 



OK, the reason for this check is that the `.ml-inference-*` indices are re-created, not updated. Consequently, it is possible to have multiple indices, but a newer index has new settings or mapped fields. 

The query that grabs the model definition objects is as follows:
```
            .setQuery(QueryBuilders.constantScoreQuery(QueryBuilders
                .boolQuery()
                .filter(QueryBuilders.termQuery(TrainedModelConfig.MODEL_ID.getPreferredName(), modelId))
                .filter(QueryBuilders.termQuery(InferenceIndexConstants.DOC_TYPE.getPreferredName(),
                    TrainedModelDefinitionDoc.NAME))))
            .setSize(MAX_NUM_DEFINITION_DOCS)
            // First find the latest index
            .addSort("_index", SortOrder.DESC)
            // Then, sort by doc_num
            .addSort(SortBuilders.fieldSort(TrainedModelDefinitionDoc.DOC_NUM.getPreferredName())
                .order(SortOrder.ASC)
                .unmappedType("long"))
            .request();
```

So, we sort by indices first and then by the doc num. 

Now, consider the situation where a model document does cross multiple indices, how do we know that the document order is guaranteed? It could be that a `doc: 0` is in the older index (index A) then a `doc: 0` is in the new index (index B), then a `doc: 1` is in the old index again. 

Now we have a broken compression string. ("doc: 0, index: B" + "doc: 0, index: A" + "doc: 1, index: B")

Preventing the index boundary cross ensures us that the documents are in the correct order. We could make this assurance in another way, but at the time, I didn't want to bother with it and thought that splitting a definition across indices was a bad idea (I still think that).

Reply:
> Preamble: I don't think configuration indices should ever be managed by ILM. The current trained model index (inference config) is not managed my ILM. This adds unnecessary complexity. Honestly, if we are THIS concerned around static model configurations causing shards to increase, we really need to come up with another storage mechanism outside the index. The more and more we talk about "large blob storage in an index" the more it gets complicated, hacky, and unmaintainable.

The problem here is that "configuration" documents are usually small and few in number.  When configuration documents are like this there is never a need for more than one index.  `.ml-config` is an example of an index like this.

When configuration documents are huge there is a risk of overflowing the 50GB recommended maximum shard size.  Arguably the documents in our `.ml-state` index are "configuration" documents, but they're not in the same index as the small documents that go in `.ml-config`.

The `.ml-inference` index is mixing the small and the large types of "configuration" in the same index.  Eventually this will result in it blowing the 50GB recommended limit in some large installation in the future.  It may be years before this happens but when it does happen it will not be easy to fix.

It may be that in the future we end up incrementing the latest inference index number not only for mappings changes, but also for scalability.

This doesn't mean that it's necessary to allow for splitting definitions across indices though.  If the index is going to be managed by bespoke ML code then that bespoke code can enforce that the models aren't split between indices.

One observation about bespoke scalability driven rollover code for complicated scenarios is that we've never got around to it in the past.  We have this problem with the ML shared results index.  For a long time there has been the idea that we might need to roll it over, or, better still, dynamically switch jobs that are generating large numbers of results to using their own dedicated results indices.  On the other hand, we _have_ managed to improve scalability for the ML state indices where the usage pattern was simple enough to just use ILM.

> The more and more we talk about "large blob storage in an index" the more it gets complicated, hacky, and unmaintainable.

Unfortunately at the moment we don't have anywhere else to put large blobs.  I agree that in the long term it would be nice if Elasticsearch had a blob storage mechanism and then we could use that for all our large opaque binary data storage needs.

Until we have have such functionality we have to use an index, and if we want to just have one index with no possibility for expansion then maybe the solution is to restrict how many models can be stored in it.  In other words, reject uploads of new models if the index would exceed a certain size afterwards.  This would force users to delete old models before uploading new ones if they were close to the limit.  It's not very "elastic", but at least protects against the index becoming unwieldy and causing problems in the cluster.

Reply:
Firstly, I don't consider it decided that we will use `.ml-inference` to store PyTorch models. It is used here because I've adapted the existing code but I've argued against it elsewhere and Dave's point about mixing small config and large blobs is a good one.

The other side of the equation is uploading the models which will play a deciding role in where the models are stored and how they are managed. I have not yet run a full end-to-end example of uploading a large model, reassembling it and performing inference, there is much to be learned from doing so. Exploring the various model upload options that have been mooted will also feedback into this design.